File: benchmarks/String/string.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

var targetSize = 1024 * 1024
var wordCount = 16
var lineLimit = 80
var fnvOffset uint64 = 14695981039346656037
var fnvPrime uint64 = 1099511628211
var referenceHash uint64 = 0xc6be9b9267a2fb8e

func randomNext ( state *uint32 ) uint32 {
	*state = *state*1664525 + 1013904223
	return *state
}


func hashBytes ( text string ) uint64 {
	hash := fnvOffset

	for i := 0; i < len(text); i++ {
		hash ^= uint64(text[i])
		hash *= fnvPrime
	}
	return hash
}


func buildSourceText ( words []string ) (string, int, uint64) {
	var text strings.Builder
	text.Grow(targetSize + 64)
	generatedWordCount := 0
	state := uint32(0x12345678)

	for {
		word := words[int(randomNext(&state)%uint32(wordCount))]
		if generatedWordCount > 0 {
			text.WriteByte(' ')
		}
		text.WriteString(word)
		generatedWordCount++
		if text.Len() > targetSize {
			break
		}
	}

	sourceText := text.String()
	return sourceText, generatedWordCount, hashBytes(sourceText)
}


func wrapWords ( words []string ) string {
	var text strings.Builder
	text.Grow(targetSize + 64)
	lineLength := 0

	for i := 0; i < len(words); i++ {
		word := words[i]
		wordLength := len(word)

		if i == 0 {
			text.WriteString(word)
			lineLength = wordLength
			continue
		}

		if lineLength+1+wordLength >= lineLimit {
			text.WriteByte('\n')
			text.WriteString(word)
			lineLength = wordLength
		} else {
			text.WriteByte(' ')
			text.WriteString(word)
			lineLength += 1 + wordLength
		}
	}

	return text.String()
}


func hashWrappedText ( text string ) uint64 {
	hash := fnvOffset
	reader := bufio.NewScanner(strings.NewReader(text))
	first := true

	for reader.Scan() {
		if !first {
			hash ^= uint64(' ')
			hash *= fnvPrime
		}
		first = false

		line := reader.Text()
		for i := 0; i < len(line); i++ {
			hash ^= uint64(line[i])
			hash *= fnvPrime
		}
	}

	return hash
}


func runBenchmark ( sourceText string, sourceHash uint64, runs int ) {
	for i := 0; i < runs; i++ {
		splitBuffer := string([]byte(sourceText))
		splitWords := strings.Split(splitBuffer, " ")
		wrappedText := wrapWords(splitWords)
		wrappedHash := hashWrappedText(wrappedText)

		if wrappedHash != sourceHash {
			fmt.Fprintf(
				os.Stderr,
				"Hash mismatch: 0x%016x != 0x%016x\n",
				wrappedHash,
				sourceHash,
			)
			os.Exit(1)
		}
	}
}


func main ( ) {
	words := []string{
		"I",
		"we",
		"cat",
		"tree",
		"café",
		"naïve",
		"jalapeño",
		"mountain",
		"blueberry",
		"basketball",
		"grandfather",
		"microbiology",
		"determination",
		"responsibility",
		"experimentation",
		"counterclockwise",
	}
	runs := 1
	if len(os.Args) > 1 {
		runs, _ = strconv.Atoi(os.Args[1])
	}
	sourceText, _, sourceHash := buildSourceText(words)

	if sourceHash != referenceHash {
		fmt.Fprintf(
			os.Stderr,
			"Source hash mismatch: 0x%016x != 0x%016x\n",
			sourceHash,
			referenceHash,
		)
		os.Exit(1)
	}

	start := time.Now()
	runBenchmark(sourceText, sourceHash, runs)
	diff := time.Since(start)
	fmt.Fprintf(os.Stderr, "Go: %.3f s\n", diff.Seconds())
}

File: benchmarks/String/string.swift
import Foundation

let targetSize = 1024 * 1024
let wordCount = 16
let lineLimit = 80
let fnvOffset: UInt64 = 14695981039346656037
let fnvPrime: UInt64 = 1099511628211
let referenceHash: UInt64 = 0xc6be9b9267a2fb8e

private
func randomNext(_ state: inout UInt32) -> UInt32
{
	state = state &* 1664525 &+ 1013904223
	return state
}


private
func hashBytes(_ text: String.UTF8View) -> UInt64
{
	var hash: UInt64 = fnvOffset

	for byte in text {
		hash ^= UInt64(byte)
		hash = hash &* fnvPrime
	}
	return hash
}


private
func buildSourceText(_ words: [String]) -> (String, Int, UInt64)
{
	var text = String()
	text.reserveCapacity(targetSize + 64)
	var length = 0
	var generatedWordCount = 0
	var state: UInt32 = 0x12345678

	while true {
		let index = Int(randomNext(&state) % UInt32(wordCount))
		let word = words[index]
		let wordLength = word.utf8.count

		if generatedWordCount > 0 {
			text.append(" ")
			length += 1
		}

		text.append(word)
		length += wordLength
		generatedWordCount += 1

		if length > targetSize {
			break
		}
	}

	return (text, generatedWordCount, hashBytes(text.utf8))
}


private
func wrapWords(_ words: [Substring]) -> String
{
	var text = String()
	text.reserveCapacity(targetSize + 64)
	var lineLength = 0

	for i in 0 ..< words.count {
		let word = words[i]
		let wordLength = word.count

		if i == 0 {
			text.append(contentsOf: word)
			lineLength = wordLength
			continue
		}

		if lineLength + 1 + wordLength >= lineLimit {
			text.append("\n")
			text.append(contentsOf: word)
			lineLength = wordLength
		} else {
			text.append(" ")
			text.append(contentsOf: word)
			lineLength += 1 + wordLength
		}
	}

	return text
}


private
func hashWrappedText(_ text: String) -> UInt64
{
	var hash: UInt64 = fnvOffset
	var first = true

	text.enumerateLines { line, _ in
		if !first {
			hash ^= 32
			hash = hash &* fnvPrime
		}
		first = false

		for byte in line.utf8 {
			hash ^= UInt64(byte)
			hash = hash &* fnvPrime
		}
	}

	return hash
}


public struct StderrOutputStream: TextOutputStream {
	public static let stream = StderrOutputStream ( )
	public func write(_ string: String) { fputs(string, stderr) }
}
var errStream = StderrOutputStream()


private
func runBenchmark(_ sourceText: String, _ sourceHash: UInt64, _ runs: Int)
{
	for _ in 0 ..< runs {
		let splitBuffer = String(
			decoding: Array(sourceText.utf8),
			as: UTF8.self
		)
		let splitWords = splitBuffer.split(separator: " ")
		let wrappedText = wrapWords(splitWords)
		let wrappedHash = hashWrappedText(wrappedText)

		if wrappedHash != sourceHash {
			print(
				String(
					format: "Hash mismatch: 0x%016llx != 0x%016llx",
					wrappedHash,
					sourceHash
				),
				to: &errStream
			)
			exit(1)
		}
	}
}


private func main ( )
{
	let words = [
		"I",
		"we",
		"cat",
		"tree",
		"café",
		"naïve",
		"jalapeño",
		"mountain",
		"blueberry",
		"basketball",
		"grandfather",
		"microbiology",
		"determination",
		"responsibility",
		"experimentation",
		"counterclockwise"
	]
	let runs = if CommandLine.arguments.count <= 1
		|| CommandLine.arguments[1] == ""
	{
		1
	} else {
		Int(CommandLine.arguments[1]) ?? 1
	}
	let (sourceText, _, sourceHash) = buildSourceText(words)

	if sourceHash != referenceHash {
		print(
			String(
				format: "Source hash mismatch: 0x%016llx != 0x%016llx",
				sourceHash,
				referenceHash
			),
			to: &errStream
		)
		exit(1)
	}

	let startTime = CFAbsoluteTimeGetCurrent()
	runBenchmark(sourceText, sourceHash, runs)
	let benchmarkTime = CFAbsoluteTimeGetCurrent() - startTime
	let timeString = String(format: "%.3f", benchmarkTime)
	print("Swift: \(timeString) s", to: &errStream)
}

main()
